lib/http/client.cpp

```diff
@@ -79,6 +88,89 @@ struct ClientPrivate
         //! All the active sessions will shutdown the connection.
         boost::signals2::signal<void()> shutdown_signal;
 };
+
+#if defined(__APPLE__)
+
+// cf_ref RAII code taken from Microsoft cpprestsdk (MIT licensed):
+// https://github.com/microsoft/cpprestsdk/blob/7fbb08c491f9c8888cc0f3d86962acb3af672772/Release/src/http/client/x509_cert_utilities.cpp#L290
+
+// Simple RAII pattern wrapper to perform CFRelease on objects.
+template<typename T>
+class cf_ref
+{
+public:
+        cf_ref(T v)
+          : value(v)
+        {
+                static_assert(sizeof(cf_ref<T>) == sizeof(T),
+                              "Code assumes just a wrapper, see usage in CFArrayCreate below.");
+        }
+        cf_ref()
+          : value(nullptr)
+        {}
+        cf_ref(cf_ref &&other)
+          : value(other.value)
+        {
+                other.value = nullptr;
+        }
+
+        ~cf_ref()
+        {
+                if (value != nullptr) {
+                        CFRelease(value);
+                }
+        }
+
+        T &get() { return value; }
+
+private:
+        cf_ref(const cf_ref &);
+        cf_ref &operator=(const cf_ref &);
+        T value;
+};
+
+bool
+import_apple_keychain(boost::asio::ssl::context &ssl_ctx_)
+{
+        cf_ref<CFArrayRef> result;
+        OSStatus osStatus;
+
+        // Copy macOS root certificates into CFArray
+        if ((osStatus = SecTrustCopyAnchorCertificates(&result.get())) != 0) {
+                std::cerr << "Error enumerating macOS certificates: " << std::endl;
+                return false;
+        }
+
+        for (CFIndex i = 0; i < CFArrayGetCount(result.get()); i++) {
+                SecCertificateRef cert = (SecCertificateRef)CFArrayGetValueAtIndex(result.get(), i);
+                cf_ref<CFDataRef> rawDataRef = SecCertificateCopyData(cert);
+                if (rawDataRef.get() == nullptr) {
+                        std::cerr << "Error enumerating macOS certificates: " << std::endl;
+                        return false;
+                }
+                const uint8_t *rawDataPtr = CFDataGetBytePtr(rawDataRef.get());
+
+                // Parse an openssl X509 object from each returned certificate
+                X509 *x509Cert = d2i_X509(NULL, &rawDataPtr, CFDataGetLength(rawDataRef.get()));
+                if (!x509Cert) {
+                        std::cerr << "Error parsing X509 Certificate from system keychain"
+                                  << std::endl;
+                        return false;
+                }
+                // Add the parsed X509 object to the X509_STORE verification store
+                const auto addStatus =
+                  X509_STORE_add_cert(SSL_CTX_get_cert_store(ssl_ctx_.native_handle()), x509Cert);
+                X509_free(x509Cert);
+                if (addStatus != 1) {
```

Comment:
We should probably handle that as we do on windows, i.e. change the windows one to also log errors, but not sure, if you should actually abort adding certs, if one of them fails?

Reply:
Yeah, that's a good point and something I didn't think about.  I'll make the status a function-level variable instead of block level and return it at the end.

Reply:
Is the return value even used? Maybe just log any certificate that fails?

Reply:
It's not currently being used, no.

Reply:
I would just not return anything then for now .-.